wd_helper.au3

```diff
@@ -2469,6 +2479,40 @@ Func _WD_DownloadFile($sURL, $sDest, $iOptions = Default)
 	Return SetError(__WD_Error($sFuncName, $iErr, $sParameters, $iExt), $iExt, $bResult)
 EndFunc   ;==>_WD_DownloadFile
 
+; #FUNCTION# ====================================================================================================================
+; Name ..........: _WD_DownloadAsBinary
+; Description ...: Download document (file or HTML content) as binary data from a given URL
+; Syntax ........: _WD_DownloadAsBinary($sURL)
+; Parameters ....: $sURL     - URL representing file to be downloaded
+; Return values .: Success - BinaryData (Download succeeded).
+;                  Failure - '' (Download failed) and sets @error to one of the following values:
+;                  - $_WD_ERROR_GeneralError
+; Author ........: mLipok
+; Modified ......:
+; Remarks .......: in case of error @extended will follow WinHttp.WinHttpRequest.5.1 @error number
+; Related .......:
+; Link ..........:
+; Example .......: No
+; ===============================================================================================================================
+Func _WD_DownloadAsBinary($sURL)
+	Local Const $sFuncName = "_WD_DownloadAsBinary"
+	Local Const $sParameters = 'Parameters:    URL=' & $sURL
+	Local $iErr = $_WD_ERROR_Success, $iExt = 0
+	Local $dResult
+
+	Local $oHTTP = ObjCreate("WinHttp.WinHttpRequest.5.1")
+	If Not @error Then $oHTTP.Open("GET", $sURL, False)
+	If Not @error Then $oHTTP.SetRequestHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0")
+	If Not @error Then $oHTTP.Send("")
+	If Not @error Then $dResult = $oHTTP.ResponseBody
```

Comment:
Each of the `@error` macros return 0. I could not force the object to return an other value.
If you can confirm this, the macros can be removed (not necessary).

Reply:
What if the given URL does not exist because, for example, it has changed?

Reply:
That is what I did, next to outer tests. I guess the object `Local $oHTTP = ObjCreate("WinHttp.WinHttpRequest.5.1")` does not provide `@error` macro syntax. Maybe you try to check some of the first characters of `$dResult` to ensure there is a response body. In case not, set error like you do.

Reply:
I tried the following, which looks bad to be honest, but it should be okay to get the problem.

<details>
<summary>wd-helper-test.au3</summary>
<p>

``` autoit
_Main()

Func _Main()
    Local Const $sURL = "https://github.com/Danp2/au3WebDriver/releases/latest"
    ;~ Local Const $sURL = "https://github.com/Danp2/au3WebDriver/releases/555latest"
    ;~ Local Const $sURL = "http://google.123.de"
    ;~ Local Const $sURL = ""

    Local $sResult = InetRead($sURL)

    ConsoleWrite(StringFormat('1:	 %s | %s\n', @error, @extended))

    If @error Then
        $sResult = _WD_DownloadAsBinary($sURL)
        ConsoleWrite(StringFormat('2:	 %s | %s\n', @error, @extended))
        ConsoleWrite(StringFormat('3:	 %s\n', $sResult))
    Else
        ConsoleWrite(StringFormat('4:	 %s | %s\n', @error, @extended))
        ;~ $sResult = _WD_DownloadAsBinary($sURL)
    EndIf
EndFunc

Func _WD_DownloadAsBinary($sURL)
    Local Const $_WD_ERROR_GeneralError = 'err1'
    Local Const $_WD_ERROR_Success = 'err2'

	Local Const $sFuncName = "_WD_DownloadAsBinary"
	Local Const $sParameters = 'Parameters:    URL=' & $sURL
	Local $iErr = $_WD_ERROR_Success, $iExt = 0
	Local $dResult

	Local $oHTTP = ObjCreate("WinHttp.WinHttpRequest.5.1")
    ConsoleWrite(StringFormat('5:	 %s | %s\n', @error, @extended))
    ConsoleWrite(StringFormat('6:	 %s\n', $oHTTP))

        $oHTTP.Open("GET", $sURL, False)
        ConsoleWrite(StringFormat('7:	 %s | %s\n', @error, @extended))

        $oHTTP.SetRequestHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0")
        ConsoleWrite(StringFormat('8:	 %s | %s\n', @error, @extended))

        $oHTTP.Send("")
        ConsoleWrite(StringFormat('9:	 %s | %s\n', @error, @extended))

        $dResult = $oHTTP.ResponseBody
        ConsoleWrite(StringFormat('10:	 %s | %s\n', @error, @extended))

    If @error Then
        ConsoleWrite(StringFormat('11:	 %s | %s\n', @error, @extended))
		$iErr = $_WD_ERROR_GeneralError
		$iExt = @error
		$dResult = ''
	EndIf

    ConsoleWrite(StringFormat('%s, %s, %s, %s\n', $sFuncName, $iErr, $sParameters, $iExt))

    Return SetError(33, 66, 0)
	;~ Return SetError(__WD_Error($sFuncName, $iErr, $sParameters, $iExt), $iExt, $dResult)
EndFunc   ;==>_WD_DownloadAsBinary

```

<p>
</details>

Reply:
my testing script:
```autoit
;~ https://github.com/Danp2/au3WebDriver/pull/512#discussion_r1508179073

Global $oErrorHandler = ObjEvent("AutoIt.Error", _ErrFunc)

_Main()

Func _Main()
    Local Const $sURL = "http://fake.url.com"
;~     Local Const $sURL = "https://github.com/Danp2/au3WebDriver/releases/latest"
;~     Local Const $sURL = "https://github.com/Danp2/au3WebDriver/releases/555latest"
;~     Local Const $sURL = ""

    Local $sResult = InetRead($sURL)

    ConsoleWrite(StringFormat('1:	 %s | %s\n', @error, @extended))

    If @error Or 1 Then ; 1 make sure that _WD_DownloadAsBinary() will be used
        $sResult = _WD_DownloadAsBinary($sURL)
        ConsoleWrite(StringFormat('2:	 %s | %s\n', @error, @extended))
        ConsoleWrite(StringFormat('3:	 %s\n', $sResult))
    Else
        ConsoleWrite(StringFormat('4:	 %s | %s\n', @error, @extended))
        ;~ $sResult = _WD_DownloadAsBinary($sURL)
    EndIf
EndFunc

Func _WD_DownloadAsBinary($sURL)
    Local Const $_WD_ERROR_GeneralError = 'err1'
    Local Const $_WD_ERROR_Success = 'err2'

	Local Const $sFuncName = "_WD_DownloadAsBinary"
	Local Const $sParameters = 'Parameters:    URL=' & $sURL
	Local $iErr = $_WD_ERROR_Success, $iExt = 0
	Local $dResult
	#forceref $dResult

	Local $oHTTP = ObjCreate("WinHttp.WinHttpRequest.5.1")
    ConsoleWrite(StringFormat('5:	 %s | %s\n', @error, @extended))
    ConsoleWrite(StringFormat('6:	 %s\n', $oHTTP))

        $oHTTP.Open("GET", $sURL, False)
        ConsoleWrite(StringFormat('7:	 %s | %s\n', @error, @extended))

        $oHTTP.SetRequestHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0")
        ConsoleWrite(StringFormat('8:	 %s | %s\n', @error, @extended))

        $oHTTP.Send("")
        ConsoleWrite(StringFormat('9:	 %s | %s\n', @error, @extended))

        $dResult = $oHTTP.ResponseBody
        ConsoleWrite(StringFormat('10:	 %s | %s\n', @error, @extended))

    If @error Then
        ConsoleWrite(StringFormat('11:	 %s | %s\n', @error, @extended))
		$iErr = $_WD_ERROR_GeneralError
		$iExt = @error
		$dResult = ''
	EndIf

    ConsoleWrite(StringFormat('%s, %s, %s, %s\n', $sFuncName, $iErr, $sParameters, $iExt))

    Return SetError(33, 66, 0)
	;~ Return SetError(__WD_Error($sFuncName, $iErr, $sParameters, $iExt), $iExt, $dResult)
EndFunc   ;==>_WD_DownloadAsBinary

; User's COM error function. Will be called if COM error occurs
Func _ErrFunc($oError)
        ; Do anything here.
        ConsoleWrite(@ScriptName & " (" & $oError.scriptline & ") : ==> COM Error intercepted !" & @CRLF & _
                        @TAB & "err.number is: " & @TAB & @TAB & "0x" & Hex($oError.number) & @CRLF & _
                        @TAB & "err.windescription:" & @TAB & $oError.windescription & @CRLF & _
                        @TAB & "err.description is: " & @TAB & $oError.description & @CRLF & _
                        @TAB & "err.source is: " & @TAB & @TAB & $oError.source & @CRLF & _
                        @TAB & "err.helpfile is: " & @TAB & $oError.helpfile & @CRLF & _
                        @TAB & "err.helpcontext is: " & @TAB & $oError.helpcontext & @CRLF & _
                        @TAB & "err.lastdllerror is: " & @TAB & $oError.lastdllerror & @CRLF & _
                        @TAB & "err.scriptline is: " & @TAB & $oError.scriptline & @CRLF & _
                        @TAB & "err.retcode is: " & @TAB & "0x" & Hex($oError.retcode) & @CRLF & @CRLF)
EndFunc   ;==>_ErrFunc
```

Reply:
take a look on:
[error handling](https://github.com/Danp2/au3WebDriver/pull/512/commits/8c54a7775f076919ab62e90e9d98384b4b3ade28)

Reply:
That's great @mlipok . I like it and I learned a lot by this COM error handling stuff 🤝 .
Thanks, go ahead with the PR.

Reply:
Take a look also here
https://github.com/Danp2/au3WebDriver/pull/513